### ISSUE TITLE
Inferenceservice no url available, ready state is False with kserve RawDeployment mode

### ISSUE THREAD

**Issue by revolutionisme** [kserve/kserve#1872]
/kind bug

**What steps did you take and what happened:**
Installed kserve with `RawDeployment`  mode, everything was installed successfully. Then followed the first inference service example from - https://kserve.github.io/website/get_started/first_isvc/
The inference service was installed without errors but the status for the `inferenceservice` always remained false and no URL is shown as can be seen below. 
![image](https://user-images.githubusercontent.com/738958/137940795-54cbf15c-e0e0-40b4-b979-0e33eeede41b.png)

We followed this similar issue which was created https://github.com/kserve/kserve/issues/1839  and stopped the `kserve-controller-manager-0` but the issue persisted. 

The deployment is in a ready state. 

**What did you expect to happen:**
The `inferenceservice` should have a url and the `ready` state should be `TRUE`

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]
We don't have istio installed as we are using nginx, also we don't have knative as well. 

**Environment:**

- Istio Version: NA
- Knative Version: NA
- KFServing Version: 0.7.0 (kserve, not kfserving)
- Kubeflow version: NA 
- Kfdef:aws
- Minikube version:NA
- Kubernetes version: (use `kubectl version`): Client - 1.22, Server(EKS) - 1.18
- OS (e.g. from `/etc/os-release`): macOS Big Sur ver 11.6

**Comment by yuzisun** [comment]
@revolutionisme What's the output of `kubectl get inferenceservices sklearn-iris -n kserve-test -oyaml` ?

**Comment by revolutionisme** [comment]
@yuzisun Please find the output below: 

```
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"serving.kserve.io/v1beta1","kind":"InferenceService","metadata":{"annotations":{},"name":"sklearn-iris","namespace":"kserve-test"},"spec":{"predictor":{"sklearn":{"storageUri":"gs://kfserving-samples/models/sklearn/iris"}}}}
  creationTimestamp: "2021-10-19T12:46:36Z"
  finalizers:
  - inferenceservice.finalizers
  generation: 1
  name: sklearn-iris
  namespace: kserve-test
  resourceVersion: "13194300"
  selfLink: /apis/serving.kserve.io/v1beta1/namespaces/kserve-test/inferenceservices/sklearn-iris
  uid: 7979c34e-2be0-4836-9328-d406c6cc7932
spec:
  predictor:
    sklearn:
      name: kserve-container
      protocolVersion: v1
      resources:
        limits:
          cpu: "1"
          memory: 2Gi
        requests:
          cpu: "1"
          memory: 2Gi
      runtimeVersion: v0.7.0
      storageUri: gs://kfserving-samples/models/sklearn/iris
status:
  components:
    predictor:
      latestCreatedRevision: "1"
  conditions:
  - lastTransitionTime: "2021-10-19T12:46:37Z"
    reason: Predictor ingress not created
    status: "False"
    type: IngressReady
  - lastTransitionTime: "2021-10-19T12:46:37Z"
    message: Deployment does not have minimum availability.
    reason: MinimumReplicasUnavailable
    status: "False"
    type: PredictorReady
  - lastTransitionTime: "2021-10-19T12:46:37Z"
    reason: Predictor ingress not created
    status: "False"
    type: Ready
```

**Comment by yuzisun** [comment]
@revolutionisme Can you help print the events with `kubectl get events -n kserve-test` ?

**Comment by yuzisun** [comment]
Also get the output of `kubectl describe pods  -l serving.kserve.io/inferenceservice=sklearn-iris -n kserve-test`

**Comment by revolutionisme** [comment]
@yuzisun Please find the output of the commands you specified. (I have just masked a couple of IP adresses with x)

```
 ~ kubectl get events -n kserve-test                                                                                                                                                               INT ✘  ml-test ⎈
No resources found in kserve-test namespace.
 ~ kubectl describe pods -l serving.kserve.io/inferenceservice=sklearn-iris -n kserve-test                                                                                                        ✔  4s  ml-test ⎈
Name:         sklearn-iris-predictor-default-6d6d7f5bdf-6gvwn
Namespace:    kserve-test
Priority:     0
Node:         ip-x-x-x-x.eu-central-1.compute.internal/x.x.x.x
Start Time:   Tue, 19 Oct 2021 14:46:37 +0200
Labels:       app=isvc.sklearn-iris-predictor-default
              component=predictor
              pod-template-hash=6d6d7f5bdf
              serving.kserve.io/inferenceservice=sklearn-iris
Annotations:  internal.serving.kserve.io/storage-initializer-sourceuri: gs://kfserving-samples/models/sklearn/iris
              kubernetes.io/psp: eks.privileged
Status:       Running
IP:           x.x.x.x
IPs:
  IP:           x.x.x.x
Controlled By:  ReplicaSet/sklearn-iris-predictor-default-6d6d7f5bdf
Init Containers:
  storage-initializer:
    Container ID:  docker://1ec214ba58546fa6c6dfa85b2a5f8ed38362e307acb8497b5fd2d9c7cb5dd13d
    Image:         kserve/storage-initializer:v0.7.0
    Image ID:      docker-pullable://kserve/storage-initializer@sha256:f92bde78bccdb6daf41d937f17f4e4eefda01bfc815c54d67d34a33fb617943e
    Port:          <none>
    Host Port:     <none>
    Args:
      gs://kfserving-samples/models/sklearn/iris
      /mnt/models
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Tue, 19 Oct 2021 14:46:38 +0200
      Finished:     Tue, 19 Oct 2021 14:46:40 +0200
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     1
      memory:  1Gi
    Requests:
      cpu:        100m
      memory:     100Mi
    Environment:  <none>
    Mounts:
      /mnt/models from kserve-provision-location (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-brsws (ro)
Containers:
  kserve-container:
    Container ID:  docker://c09706b1a7279c98ab3dbdcb4a770069be129bdeb3e94a29366422af5880dc84
    Image:         kserve/sklearnserver:v0.7.0
    Image ID:      docker-pullable://kserve/sklearnserver@sha256:0849f9c43c716e85b198bde993bb155a3f65dac6568d0d5565b56b99fb08dcb3
    Port:          <none>
    Host Port:     <none>
    Args:
      --model_name=sklearn-iris
      --model_dir=/mnt/models
      --http_port=8080
    State:          Running
      Started:      Tue, 19 Oct 2021 14:46:40 +0200
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     1
      memory:  2Gi
    Requests:
      cpu:        1
      memory:     2Gi
    Readiness:    tcp-socket :8080 delay=0s timeout=1s period=10s #success=1 #failure=3
    Environment:  <none>
    Mounts:
      /mnt/models from kserve-provision-location (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-brsws (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             True
  ContainersReady   True
  PodScheduled      True
Volumes:
  default-token-brsws:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-brsws
    Optional:    false
  kserve-provision-location:
    Type:        EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:
    SizeLimit:   <unset>
QoS Class:       Burstable
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                 node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:          <none>
```

**Comment by revolutionisme** [comment]
@yuzisun  Another thing is that we have the deployment and the service running (Have described them both)

```
 ~ kubectl describe deployments -n kserve-test                                                                                                                                                                       1 ✘  ml-test ⎈
Name:                   sklearn-iris-predictor-default
Namespace:              kserve-test
CreationTimestamp:      Tue, 19 Oct 2021 14:46:37 +0200
Labels:                 app=isvc.sklearn-iris-predictor-default
                        component=predictor
                        serving.kserve.io/inferenceservice=sklearn-iris
Annotations:            deployment.kubernetes.io/revision: 1
                        internal.serving.kserve.io/storage-initializer-sourceuri: gs://kfserving-samples/models/sklearn/iris
Selector:               app=isvc.sklearn-iris-predictor-default
Replicas:               1 desired | 1 updated | 1 total | 1 available | 0 unavailable
StrategyType:           RollingUpdate
MinReadySeconds:        0
RollingUpdateStrategy:  25% max unavailable, 25% max surge
Pod Template:
  Labels:       app=isvc.sklearn-iris-predictor-default
                component=predictor
                serving.kserve.io/inferenceservice=sklearn-iris
  Annotations:  internal.serving.kserve.io/storage-initializer-sourceuri: gs://kfserving-samples/models/sklearn/iris
  Containers:
   kserve-container:
    Image:      kserve/sklearnserver:v0.7.0
    Port:       <none>
    Host Port:  <none>
    Args:
      --model_name=sklearn-iris
      --model_dir=/mnt/models
      --http_port=8080
    Limits:
      cpu:     1
      memory:  2Gi
    Requests:
      cpu:        1
      memory:     2Gi
    Readiness:    tcp-socket :8080 delay=0s timeout=1s period=10s #success=1 #failure=3
    Environment:  <none>
    Mounts:       <none>
  Volumes:        <none>
Conditions:
  Type           Status  Reason
  ----           ------  ------
  Available      True    MinimumReplicasAvailable
  Progressing    True    NewReplicaSetAvailable
OldReplicaSets:  <none>
NewReplicaSet:   sklearn-iris-predictor-default-6d6d7f5bdf (1/1 replicas created)
Events:          <none>
```

```                                                                                                                                                                 
 ~ kubectl describe services -n kserve-test                                                                                                                                                                            ✔  ml-test ⎈
Name:              sklearn-iris-predictor-default
Namespace:         kserve-test
Labels:            app=isvc.sklearn-iris-predictor-default
                   component=predictor
                   serving.kserve.io/inferenceservice=sklearn-iris
Annotations:       internal.serving.kserve.io/storage-initializer-sourceuri: gs://kfserving-samples/models/sklearn/iris
Selector:          app=isvc.sklearn-iris-predictor-default
Type:              ClusterIP
IP Families:       <none>
IP:                x.x.x.x
IPs:               <none>
Port:              sklearn-iris-predictor-default  80/TCP
TargetPort:        8080/TCP
Endpoints:         x.x.x.x:8080
Session Affinity:  None
Events:            <none>
```

**Comment by yuzisun** [comment]
@revolutionisme Which ingress you are using ? If you can get the kserve controller logs from kserve namespace that should give us clue.

**Comment by revolutionisme** [comment]
@yuzisun So we have 2 things, 

1. Regarding the output of the `kubectl get events -n kserve-test`, we deleted the inferenceservice and recreated it and now we have some logs for the same.

```
 ~  kubectl get events -n kserve-test                                                                                                       ✔ ml-test ⎈
LAST SEEN   TYPE      REASON                         OBJECT                                                   MESSAGE
114s        Normal    Killing                        pod/sklearn-iris-predictor-default-6d6d7f5bdf-6gvwn      Stopping container kserve-container
86s         Normal    Scheduled                      pod/sklearn-iris-predictor-default-6d6d7f5bdf-v2slq      Successfully assigned kserve-test/sklearn-iris-predictor-default-6d6d7f5bdf-v2slq to ip-10-6-38-226.eu-central-1.compute.internal
85s         Normal    Pulled                         pod/sklearn-iris-predictor-default-6d6d7f5bdf-v2slq      Container image "kserve/storage-initializer:v0.7.0" already present on machine
85s         Normal    Created                        pod/sklearn-iris-predictor-default-6d6d7f5bdf-v2slq      Created container storage-initializer
85s         Normal    Started                        pod/sklearn-iris-predictor-default-6d6d7f5bdf-v2slq      Started container storage-initializer
83s         Normal    Pulled                         pod/sklearn-iris-predictor-default-6d6d7f5bdf-v2slq      Container image "kserve/sklearnserver:v0.7.0" already present on machine
82s         Normal    Created                        pod/sklearn-iris-predictor-default-6d6d7f5bdf-v2slq      Created container kserve-container
82s         Normal    Started                        pod/sklearn-iris-predictor-default-6d6d7f5bdf-v2slq      Started container kserve-container
86s         Normal    SuccessfulCreate               replicaset/sklearn-iris-predictor-default-6d6d7f5bdf     Created pod: sklearn-iris-predictor-default-6d6d7f5bdf-v2slq
86s         Normal    ScalingReplicaSet              deployment/sklearn-iris-predictor-default                Scaled up replica set sklearn-iris-predictor-default-6d6d7f5bdf to 1
70s         Warning   FailedGetResourceMetric        horizontalpodautoscaler/sklearn-iris-predictor-default   unable to get metrics for resource cpu: no metrics returned from resource metrics API
70s         Warning   FailedComputeMetricsReplicas   horizontalpodautoscaler/sklearn-iris-predictor-default   invalid metrics (1 invalid out of 1), first error is: failed to get cpu utilization: unable to get metrics for resource cpu: no metrics returned from resource metrics API
22s         Warning   FailedGetResourceMetric        horizontalpodautoscaler/sklearn-iris-predictor-default   did not receive metrics for any ready pods
22s         Warning   FailedComputeMetricsReplicas   horizontalpodautoscaler/sklearn-iris-predictor-default   invalid metrics (1 invalid out of 1), first error is: failed to get cpu utilization: did not receive metrics for any ready pods
86s         Warning   UpdateFailed                   inferenceservice/sklearn-iris                            Failed to update status for InferenceService "sklearn-iris": Operation cannot be fulfilled on inferenceservices.serving.kserve.io "sklearn-iris": the object has been modified; please apply your changes to the latest version and try again
86s         Warning   InternalError                  inferenceservice/sklearn-iris                            fails to update InferenceService status: Operation cannot be fulfilled on inferenceservices.serving.kserve.io "sklearn-iris": the object has been modified; please apply your changes to the latest version and try again
```

2. We found out the following error logs after you asked us to check the kserve controller logs . They are as follows:

```
│ manager {"level":"error","ts":1634739997.7795963,"logger":"controller-runtime.manager.controller.inferenceservice","msg":"Reconciler error","reconciler group":"serving.kserve.io","reconciler kind":"InferenceService","name":"sklearn-ir │
│ is","namespace":"kserve-test","error":"fails to reconcile ingress: no matches for kind \"Ingress\" in version \"networking.k8s.io/v1\"","errorVerbose":"no matches for kind \"Ingress\" in version \"networking.k8s.io/v1\"\nfails to reco │
│ ncile ingress\ngithub.com/kserve/kserve/pkg/controller/v1beta1/inferenceservice.(*InferenceServiceReconciler).Reconcile\n\t/go/src/github.com/kserve/kserve/pkg/controller/v1beta1/inferenceservice/controller.go:181\nsigs.k8s.io/control │
│ ler-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:263\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).pr │
│ ocessNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:235\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.1\n\t/go/pkg/mod/sigs.k8s.io/controller │
│ -runtime@v0.7.0/pkg/internal/controller/controller.go:198\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.BackoffU │
│ ntil.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.Ji │
│ tterUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext\n\t/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/uti │
│ l/wait.UntilWithContext\n\t/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:99\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1373","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/pkg/mod/github. │
│ com/go-logr/zapr@v0.2.0/zapr.go:132\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:267\nsigs.k8s.io/con │
│ troller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:235\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Control │
│ ler).Start.func1.1\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.0/pkg/internal/controller/controller.go:198\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait │
│ /wait.go:185\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:155\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/uti │
│ l/wait/wait.go:156\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext\n\t/go/pkg/mod/k8s.io/apimachinery@v0.19. │
│ 2/pkg/util/wait/wait.go:185\nk8s.io/apimachinery/pkg/util/wait.UntilWithContext\n\t/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:99"}
```

This seems related to this issue on kubernetes - https://github.com/kubernetes/kubernetes/issues/90077 as we have Kubernetes 1.18 (On EKS) as previously mentioned and the ingress for networking.k8s.io/v1 is available from 1.19. Can you confirm? If yes, is there a way to configure ingress with networking.k8s.io/v1beta1 for kserve?

**Comment by Iamlovingit** [comment]
As @revolutionisme mentioned, `networking.k8s.io/v1` is available from kubernetes 1.19, 1.18 is not supported. there is the link:https://kubernetes.io/docs/concepts/services-networking/ingress/ 
There are two ways to fix this:
1. update your k8s 
2. build a specific version kserve version in your cluster, change the ingress api to old one. 
**If you want to build your version, I'm happy to help you.**

**Comment by revolutionisme** [comment]
@Iamlovingit We would love your help building a custom version for our k8s as updating the cluster wouldn't be possible for now. How can we coordinate this?

**Comment by Iamlovingit** [comment]
Hi @revolutionisme I think you just need to change this line.
https://github.com/kserve/kserve/blob/fb239bde6395e107a9e884ff36671cdf23a0571d/pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/kube_ingress_reconciler.go#L24

```golang
// netv1 "k8s.io/api/networking/v1" 
netv1 "k8s.io/api/networking/v1beta1" 
```

then follow the doc to deploy your version:
https://github.com/kserve/website/blob/main/docs/developer/developer.md#deploy-kserve-with-your-own-version

If you have any question, you can contact me by slack in kfserving channel.

**Comment by revolutionisme** [comment]
Thanks a lot @Iamlovingit for all the help, I am just pasting all the instructions here for anyone who wants to dpeloy kserve on kubernetes 1.18 with `RawDeployement` mode (As networking v1 is not available in 1.18)

Follow all the instructions above, then follow the steps below:

1. Before you run make deploy-dev, you should change the default deployment to RawDeployment here `"defaultDeploymentMode": "RawDeployment"` https://github.com/kserve/kserve/blob/fb239bde6395e107a9e884ff36671cdf23a0571d/config/configmap/inferenceservice.yaml#L190
2. In the `github.com/kserve/kserve/pkg/controller/v1beta1/inferenceservice/reconcilers/ingress
pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/kube_ingress_reconciler.go:73` change the ingress backend structure to 

```
Backend: netv1.IngressBackend{
                            ServiceName: componentName,
                            ServicePort: intstr.FromInt(constants.CommonDefaultHttpPort),
                        },
```

Then run ` make dev-deploy`, and use the context of the Kubernetes cluster where you want to deploy this. Also, you may have to stop the `kserve-controller-manager-0` to get the inference URL as mentioned in this ticket here - https://github.com/kserve/kserve/issues/1839
